mlir/include/mlir/IR/BuiltinTypes.td

```diff
@@ -40,6 +40,11 @@ def ValueSemantics : NativeTypeTrait<"ValueSemantics"> {
   let cppNamespace = "::mlir";
 }
 
+/// Type trait indicating that the type is a pointer-like type.
```

Comment:
This would deserve a bit more documentation: what's the implication of marking a type with "pointer-like"?
Why is memref not a a PointerLike type for example?

Reply:
Note: we wouldn't have to answer this kind of difficult question if we were making `Builtin_VectorTypeElementType` a TypeInferface instead, we could implement it on the in-tree "pointer-like" types we select.

Reply:
What if we rename it to `BarePointerLike`?